cmd/segment-reaper/detect.go

```diff
@@ -25,9 +35,126 @@ var (
 	}
 )
 
+// Cluster key for objects map
+type Cluster struct {
+	projectID string
+	bucket    string
+}
+
+// Object represents object with segments
+type Object struct {
+	// TODO verify if we have more than 64 segments for object in network
+	segments uint64
+
+	expectedNumberOfSegments byte
+
+	hasLastSegment bool
+	// if skip is true segments from object should be removed from memory when last segment is found
+	// or iteration is finished, mark it as true if one of the segments from this object is newer
+	// then specified threshold
+	skip bool
+}
+
+// ObjectsMap map that keeps objects representation
+type ObjectsMap map[Cluster]map[storj.Path]*Object
+
+// Observer metainfo.Loop observer for zombie reaper
+type Observer struct {
+	db      metainfo.PointerDB
+	objects ObjectsMap
+	writer  *csv.Writer
+
+	lastProjectID string
+
+	inlineSegments     int
+	lastInlineSegments int
+	remoteSegments     int
+}
+
+// RemoteSegment process segment to collect data needed to detect zombie segment
+func (observer *Observer) RemoteSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	return observer.processSegment(ctx, path, pointer)
+}
+
+// InlineSegment process segment to collect data needed to detect zombie segment
+func (observer *Observer) InlineSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	return observer.processSegment(ctx, path, pointer)
+}
+
+// Object not used in this implementation
+func (observer *Observer) Object(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) (err error) {
+	return nil
+}
+
+func (observer *Observer) processSegment(ctx context.Context, path metainfo.ScopedPath, pointer *pb.Pointer) error {
+	cluster := Cluster{
+		projectID: path.ProjectIDString,
+		bucket:    path.BucketName,
+	}
+
+	object := findOrCreate(cluster, path.EncryptedObjectPath, observer.objects)
+	if observer.lastProjectID != "" && observer.lastProjectID != cluster.projectID {
+		err := analyzeProject(ctx, observer.db, observer.objects, observer.writer)
+		if err != nil {
+			return err
+		}
+
+		// cleanup map to free memory
+		observer.objects = make(ObjectsMap)
+	}
+
+	isLastSegment := path.Segment == "l"
+	if isLastSegment {
+		object.hasLastSegment = true
+
+		streamMeta := pb.StreamMeta{}
+		err := proto.Unmarshal(pointer.Metadata, &streamMeta)
+		if err != nil {
+			return errs.New("unexpected error unmarshalling pointer metadata %s", err)
+		}
+
+		if streamMeta.NumberOfSegments > 0 {
+			if streamMeta.NumberOfSegments > int64(maxNumOfSegments) {
+				object.skip = true
+				zap.S().Warn("unsupported number of segments", zap.Int64("index", streamMeta.NumberOfSegments))
+			}
+			object.expectedNumberOfSegments = byte(streamMeta.NumberOfSegments)
+		}
+	} else {
+		segmentIndex, err := strconv.Atoi(path.Segment[1:])
+		if err != nil {
+			return err
+		}
+		if segmentIndex >= int(maxNumOfSegments) {
+			object.skip = true
```

Comment:
According the `skip` description the segments should be removed from memory, where is this happening?

Reply:
comment was out of date, fixed